### PR TITLE
ftfy version 4.0

### PR DESCRIPTION
**Pull request by rspeer**
- Defaults to NFC normalization instead of NFKC
- Fixes more weird stuff
- Fewer false positives (0 false positives in 13 million tweets)
- More useful command-line tool
- Knows the Unicode 7 character classes
- Warns people about eventually dropping Python 2 support

### CODE REVIEW

File: ftfy/badness.py [hunk 1]
```diff
@@ -129,7 +130,7 @@ def sequence_weirdness(text):
     """
     text2 = unicodedata.normalize('NFC', text)
     weirdness = len(WEIRDNESS_RE.findall(chars_to_classes(text2)))
-    punct_discount = len(ENDING_PUNCT_RE.findall(text2))
+    punct_discount = sum(len(x) for x in ENDING_PUNCT_RE.findall(text2))
```
Comment (akenney) [inline ftfy/badness.py]: The comment on ENDING_PUNCT_RE says "The match ends with a + so that we only give the bonus once for a consecutive sequence of these characters."  Isn't this change just undoing that?

File: ftfy/bad_codecs/utf8_variants.py [hunk 2]
```diff
@@ -146,7 +145,7 @@ def _buffer_decode_step(self, input, errors, final):
             # Set final=True because 0xc0 and 0xed don't make sense in the
             # middle of a sequence, in any variant.
             return sup(input[:cutoff], errors, True)
-
+
```
Comment (akenney) [inline ftfy/bad_codecs/utf8_variants.py]: =\

File: ftfy/cli.py [hunk 3]
```diff
@@ -17,17 +37,62 @@ def main():
...
+                        help="Ask ftfy to guess the encoding of your input. "
+                             "This is risky. Overrides -e.")
+    parser.add_argument('-e', '--encoding', type=str, default='utf-8',
+                        help='The encoding of the input. Defaults to UTF-8.')
+    parser.add_argument('-n', '--normalization', type=str, default='NFC',
+                        help='The normalization of Unicode to apply. '
+                             'Defaults to NFC. Can be "none".')
+    parser.add_argument('--preserve-entities', action='store_true',
+                        help="Leave HTML entities as they are. The default "
+                             "is to decode them in non-HTML files.")
```
Comment (akenney) [inline ftfy/cli.py]: This implies that it's looking at the file extension or the file structure or something (which in this case it probably could do and might be a good idea), but really it's still just looking for angle-brackets, right?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,39 @@
...
+  option.
+- The command-line tool can be configured with options, and can be used as a
+  pipe.
+- Recognizes characters that are new in Unicode 7.0.
+
+Deprecations:
+
+- `fix_text_encoding` has been renamed to simply `fix_encoding`, and the old
+  name emits a warning. It will disappear in ftfy 4.2 or 5.0.
+- `remove_unsafe_private_use` has been removed entirely, after two versions
```
Comment (alin-luminoso) [inline CHANGELOG.md]: Drive-by review note: this is a breaking change, not a deprecation, right?

File: ftfy/cli.py [hunk 5]
```diff
@@ -4,10 +4,30 @@
 Because files do not come with their encoding marked, it first runs the file
 through `ftfy.guess_bytes`, then runs it through `ftfy.fix_text`.
```
Comment (akenney) [inline ftfy/cli.py]: This docstring is not fully accurate now that you're supposed to tell it the encoding.

File: ftfy/fixes.py [hunk 6]
```diff
@@ -82,42 +72,71 @@ def fix_text_encoding(text):
...
+
+
 def fix_encoding_and_explain(text):
     """
     Re-decodes text that has been decoded incorrectly, and also return a
     "plan" indicating all the steps required to fix it.
 
     To fix similar text in the same way, without having to detect anything,
-    you can use the ``apply_plan`` function.
+    you can use the :func:`ftfy.fixes.apply_plan` function.
```
Comment (akenney) [inline ftfy/fixes.py]: I think what this is trying to say is that if you have a bunch of different texts that you know are all broken in the same way, you can run this function on one of them to get the "plan", and then apply that plan to all the other texts without having to re-derive it?  I think this docstring sentence could be clearer though, about what it means by "similar text".

File: ftfy/fixes.py [hunk 7]
```diff
@@ -230,20 +261,28 @@ def apply_plan(text, plan):
...
 
-    Because only text can be encoded, and only bytes can be decoded, the plan
-    should alternate 'encode' and 'decode' steps, or else this function will
-    encounter an error.
+    - `operation` is 'encode' if it turns a string into bytes, 'decode' if it
+      turns bytes into a string, and 'transcode' if it keeps the type the same.
+    - `encoding` is the name of the encoding to use, such as 'utf-8' or
+      'latin-1', or the function name in the case of 'transcode'.
+    - `cost` is a penalty to apply to the score of the resulting text, when
+      performing a dubious operation that requires a lot of evidence.
```
Comment (akenney) [inline ftfy/fixes.py]: The description of `cost` is kind of confusing, because cost is pretty much an internal thing that the user doesn't usually need to know about.  This function doesn't actually *use* the costs, but this docstring makes it sound like it might use them, or it's unclear who's doing the applying of the penalty.  Maybe you could explicitly say that `cost` is a thing used by other users of plans, namely `fix_encoding_and_explain`, which has to decide *which* plan to use.

File: docs/index.rst [hunk 8]
```diff
@@ -110,53 +194,94 @@ A note on encoding detection
...
 --------
 .. include:: accuracy.rst
 
+
+Command-line tool
+-----------------
+ftfy can be used from the command line. By default, it takes UTF-8 input and
+writes it to UTF-8 output, fixing problems in its Unicode as it goes.
+
+Here's the usage documentation for the `ftfy` command::
```
Comment (akenney) [inline docs/index.rst]: This is out of date now (there's a `-o` option and the help text for `--preserve-entities` has changed).  There's no way to auto-generate this, is there?

File: ftfy/cli.py [hunk 9]
```diff
@@ -1,33 +1,116 @@
...
+    encoding = args.encoding
+    if args.guess:
+        encoding = None
+
+    if args.filename == '-':
+        file = sys.stdin
+    else:
+        file = open(args.filename, 'rb')
+
+    encode_output = ENCODE_STDIO or (args.output != '-')
```
Comment (akenney) [inline ftfy/cli.py]: I think this line is now not used (`encode_output` gets set in the next few lines).

File: ftfy/__init__.py [hunk 10]
```diff
@@ -161,65 +186,73 @@ def fix_text(text,
     return ''.join(out)
 
 ftfy = fix_text
+fix_encoding = fixes.fix_encoding
```
Comment (akenney) [inline ftfy/__init__.py]: Sooo ... the `fix_encoding` function is being set globally in this file, but is not actually used in this file (I guess it's here so people can import it from the top level?).  Its deprecated name `fix_text_encoding` is imported directly, but is not actually used anywhere (it was previously used in the even-more-deprecated `fix_bad_encoding`); elsewhere in this file, `fix_text_encoding` is referenced as `fixes.fix_text_encoding`.  Additionally, three of the functions defined in this file take an *argument* called `fix_encoding` (not to be confused with this global variable of the same name).

This must be at least a little bit simplifiable.

File: ftfy/__init__.py [hunk 11]
```diff
@@ -18,29 +18,33 @@
 import unicodedata
 import warnings
```
Comment (akenney) [inline ftfy/__init__.py]: This import seems to be unused.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,39 @@
...
+- New heuristics, so that ftfy can fix more stuff, while maintaining
+  approximately zero false positives.
+- The command-line tool trusts you to know what encoding your *input* is in,
+  and assumes UTF-8 by default. You can still tell it to guess with the `-g`
+  option.
+- The command-line tool can be configured with options, and can be used as a
+  pipe.
+- Recognizes characters that are new in Unicode 7.0.
+- `remove_unsafe_private_use` has been removed entirely, after two versions
+  of deprecation.
```
Comment (akenney) [inline CHANGELOG.md]: The deprecated `fix_bad_encoding` has also been removed.  (Also, removal of deprecated things might be better categorized under "breaking changes" than "new features".)

File: ftfy/__init__.py (hunk 10 above)
Comment (rspeer) [inline ftfy/__init__.py]: Do you think it's okay for `fix_encoding` the parameter and `fix_encoding` the top-level function to have the same name, if I fix the other inconsistencies?

I probably called it `fix_text_encoding` before in part to avoid this parameter collision, but I want the function to have a short, accessible name because some use cases should just go straight to `fix_encoding`, not `fix_text`.

File: ftfy/streamtester/__init__.py [hunk 13]
```diff
@@ -17,14 +18,18 @@ def __init__(self):
...
         would change it. If so, display the change.
         """
         self.count += 1
+        text = unescape_html(text)
         if not possible_encoding(text, 'ascii'):
-            fixed = fix_text_encoding(text)
+            if encoding_only:
+                fixed = fix_encoding(text)
+            else:
+                fixed = fix_text(text, fix_character_width=False, uncurl_quotes=False)
```
Comment (akenney) [inline ftfy/streamtester/__init__.py]: This is for that new stuff in accuracy.rst?  It might make sense for check_ftfy to just take kwargs to pass to fix_text, but if you're the only one who uses it you're probably just going to be manually changing this stuff anyway so it probably doesn't matter?

Comment (rspeer) [inline ftfy/streamtester/__init__.py]: Yep, around now I'm manually changing it and I'm fine with that.

**Comment by alin-luminoso** [comment]
(this is only a test comment)
